src/Control/Monad/Catch.hs

```diff
@@ -653,44 +783,56 @@ catches a hs = a `catch` handler
 
 -- | Run an action only if an exception is thrown in the main action. The
 -- exception is not caught, simply rethrown.
+--
+-- /NOTE/ The action is only run if an /exception/ is throw. If the monad
+-- supports other ways of aborting the computation, the action won't run if
+-- those other kinds of errors are thrown. See 'onError'.
 onException :: MonadCatch m => m a -> m b -> m a
 onException action handler = action `catchAll` \e -> handler >> throwM e
 
+-- | Run an action only if an error is thrown in the main action. Unlike
```

Comment:
It may be worth spelling out that the difference is for transformers like `EitherT` and `MaybeT` that have multiple exit points.